File: second course/PDOO/civitasJava/src/civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;

class MazoSorpresas{
    
  private Boolean barajada;   
  private int usadas;   
  private Boolean debug;       
  private Sorpresa ultimaSorpresa;  

  //Atributos de referencia
  private ArrayList <Sorpresa> sorpresas;  
  private ArrayList <Sorpresa> cartasEspeciales;
  
  
  MazoSorpresas(Boolean d){ //TESTED
    init();
		debug = d;
	
		if(debug)
			Diario.getInstance().ocurreEvento("Modo debug del mazo: ACTIVADO.");
	}
  
  MazoSorpresas(){ //TESTED
    init();
		debug = false;
  }
  
  private void init(){
    sorpresas = new ArrayList<>();
		cartasEspeciales = new ArrayList<>();
		barajada = false;
		usadas = 0;
  }
    
  void alMazo(Sorpresa s){ //TESTED
		if(!barajada)
			sorpresas.add(s);
  }
  
  Sorpresa siguiente(){ //TESTED
		if(!barajada || usadas == sorpresas.size() + cartasEspeciales.size()){
			if(!debug){ 
				Collections.shuffle(sorpresas);
				barajada = true;
			}
			
			usadas = 0;
		}
	
		++usadas;
		ultimaSorpresa = sorpresas.remove(0);
		sorpresas.add(ultimaSorpresa);
	
		return ultimaSorpresa;
  } 
  
  void inhabilitarCartaEspecial(Sorpresa sorpresa){ //TESTED
		boolean encontrada = sorpresas.remove(sorpresa);
	
		if(encontrada){
			cartasEspeciales.add(sorpresa);
			Diario.getInstance().ocurreEvento("Carta Especial inhabilitada!!");
		}
  }
  
  void habilitarCartaEspecial(Sorpresa sorpresa){ //TESTED
		boolean encontrada = cartasEspeciales.remove(sorpresa);
	
		if(encontrada){
			sorpresas.add(sorpresa);
			Diario.getInstance().ocurreEvento("Carta Especial habilitada!!");
		}
    
  }
	
	public static void main(String [] args){
		
		MazoSorpresas mazo = new MazoSorpresas(true);
		
		mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, 0, "IRCASILLA"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, 37, "PAGARCOBRAR"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL,60, "PORCASAHOTEL"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, 50, "PORJUGADOR"));
		
		Sorpresa s;
		
		while(mazo.sorpresas.size() > 0){
			
			s = mazo.siguiente();
			System.out.println(s.toString());
			mazo.inhabilitarCartaEspecial(s);
		}
		
		System.out.println("---------");
		
		while(mazo.cartasEspeciales.size() > 0){
			
			s = mazo.cartasEspeciales.get(0);
			System.out.println(s.toString());
			mazo.habilitarCartaEspecial(s);
		}
		
		System.out.println("---------");
		
		while(Diario.getInstance().eventosPendientes())
			System.out.println(Diario.getInstance().leerEvento());
	}
  
}  
  
  
  

File: second course/PDOO/civitasRuby/pkg/civitasRuby-0.0.1/lib/mazo_sorpresas.rb
# encoding:utf-8

require_relative "sorpresa"
require_relative "diario"

module Civitas
	class Mazo_sorpresas
		
		def initialize(d = false)
			init()
			@debug = d
			
			if(@debug)
				Diario.instance.ocurre_evento("Modo debug del mazo: ACTIVADO.")
			end
		end
		
		
		private
		
		def init()
			@sorpresas = Array.new() 
			@cartas_especiales = Array.new() 
			@barajada = false
			@usadas = 0
		end
		
		
		public
		
		def al_mazo(s)
			if(!@barajada)
				@sorpresas << s
			end
		end
		
		def siguiente()
			if(!@barajada || @usadas == @sorpresas.size() + @cartas_especiales.size())
				if(!@debug)
					@sorpresas.shuffle!
					@barajada = true
				end
				@usadas = 0
			end
			
			@usadas += 1
			@ultima_sorpresa = @sorpresas.shift()
			@sorpresas << @ultima_sorpresa
			
			return @ultima_sorpresa
		end
		
		def inhabilitar_carta_especial(sorpresa)
			for i in (0 .. @sorpresas.size() - 1)
        if(@sorpresas[i] == sorpresa)
          aux = @sorpresas[i]
          @sorpresas.delete_at(i)
          @cartas_especiales << aux
          Diario.instance.ocurre_evento("Carta Especial inhabilitada!!")
        end
      end
		end
		
		def habilitar_carta_especial(sorpresa)
    
      for i in (0 .. @cartas_especiales.size() - 1)
        if(@cartas_especiales[i] == sorpresa)
          aux = @cartas_especiales[i]
          @cartas_especiales.delete_at(i)
          @sorpresas << aux
          Diario.instance.ocurre_evento("Carta Especial habilitada!!")
        end
      end
    end
		
		
	end
end
